pinot-core/src/main/java/com/linkedin/pinot/core/common/DataFetcher.java

```diff
@@ -0,0 +1,205 @@
+/**
+ * Copyright (C) 2014-2015 LinkedIn Corp. (pinot-core@linkedin.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.core.common;
+
+import com.linkedin.pinot.core.indexsegment.IndexSegment;
+import com.linkedin.pinot.core.segment.index.readers.Dictionary;
+import java.util.HashMap;
+import java.util.Map;
+
+
+/**
+ * DataFetcher is a higher level abstraction for data fetching. Given an index segment, DataFetcher can manage the
+ * DataSource, Dictionary, BlockValSet and BlockValIterator for this segment, preventing redundant construction for
+ * these instances. DataFetcher can be used by both selection, aggregation and group-by data fetching process, reducing
+ * duplicate codes and garbage collection.
+ */
+public class DataFetcher {
+  private final IndexSegment _indexSegment;
+
+  private final Map<String, DataSource> _columnToDataSourceMap = new HashMap<>();
+  private final Map<String, Dictionary> _columnToDictionaryMap = new HashMap<>();
+  private final Map<String, BlockValSet> _columnToBlockValSetMap = new HashMap<>();
+
+  /**
+   * Constructor for DataFetcher.
+   *
+   * @param indexSegment index segment.
+   */
+  public DataFetcher(IndexSegment indexSegment) {
+    _indexSegment = indexSegment;
+  }
+
+  /**
+   * Given a column, fetch its data source.
+   *
+   * @param column column name.
+   * @return data source associated with this column.
+   */
+  public DataSource getDataSourceForColumn(String column) {
```

Comment:
This class manages the contents inside IndexSegment. For some use case, we want to get the metadata for a column, with this class, we don't need to create a separate DataSource each time, and can reduce the garbage. For multi-value column, we can get the blockValSet and use it to fetch data. I didn't implement a fetchMultiData function because that will add a map look up for each docId, which is not efficient.